staging/src/k8s.io/apiserver/pkg/cel/openapi/values.go

```diff
@@ -0,0 +1,702 @@
+/*
+Copyright 2021 The Kubernetes Authors.
```

Comment:
(since git doesn't do a great job of showing the file moved and was largely unchanged)

```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package cel
+package openapi
 
 import (
 	"fmt"
@@ -26,24 +26,23 @@ import (
 	"github.com/google/cel-go/common/types/ref"
 	"github.com/google/cel-go/common/types/traits"
 
-	structuralschema "k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
-	"k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/model"
 	"k8s.io/apimachinery/pkg/api/equality"
 	"k8s.io/apiserver/pkg/cel"
+	"k8s.io/kube-openapi/pkg/validation/spec"
 	"k8s.io/kube-openapi/pkg/validation/strfmt"
 )
 
 // UnstructuredToVal converts a Kubernetes unstructured data element to a CEL Val.
 // The root schema of custom resource schema is expected contain type meta and object meta schemas.
 // If Embedded resources do not contain type meta and object meta schemas, they will be added automatically.
-func UnstructuredToVal(unstructured interface{}, schema *structuralschema.Structural) ref.Val {
+func UnstructuredToVal(unstructured interface{}, schema *spec.Schema) ref.Val {
 	if unstructured == nil {
 		if schema.Nullable {
 			return types.NullValue
 		}
 		return types.NewErr("invalid data, got null for schema with nullable=false")
 	}
-	if schema.XIntOrString {
+	if isXIntOrString(schema) {
 		switch v := unstructured.(type) {
 		case string:
 			return types.String(v)
@@ -56,19 +55,19 @@ func UnstructuredToVal(unstructured interface{}, schema *structuralschema.Struct
 		}
 		return types.NewErr("invalid data, expected XIntOrString value to be either a string or integer")
 	}
-	if schema.Type == "object" {
+	if schema.Type.Contains("object") {
 		m, ok := unstructured.(map[string]interface{})
 		if !ok {
 			return types.NewErr("invalid data, expected a map for the provided schema with type=object")
 		}
-		if schema.XEmbeddedResource || schema.Properties != nil {
-			if schema.XEmbeddedResource {
-				schema = model.WithTypeAndObjectMeta(schema)
+		if isXEmbeddedResource(schema) || schema.Properties != nil {
+			if isXEmbeddedResource(schema) {
+				schema = WithTypeAndObjectMeta(schema)
 			}
 			return &unstructuredMap{
 				value:  m,
 				schema: schema,
-				propSchema: func(key string) (*structuralschema.Structural, bool) {
+				propSchema: func(key string) (*spec.Schema, bool) {
 					if schema, ok := schema.Properties[key]; ok {
 						return &schema, true
 					}
@@ -76,22 +75,22 @@ func UnstructuredToVal(unstructured interface{}, schema *structuralschema.Struct
 				},
 			}
 		}
-		if schema.AdditionalProperties != nil && schema.AdditionalProperties.Structural != nil {
+		if schema.AdditionalProperties != nil && schema.AdditionalProperties.Schema != nil {
 			return &unstructuredMap{
 				value:  m,
 				schema: schema,
-				propSchema: func(key string) (*structuralschema.Structural, bool) {
-					return schema.AdditionalProperties.Structural, true
+				propSchema: func(key string) (*spec.Schema, bool) {
+					return schema.AdditionalProperties.Schema, true
 				},
 			}
 		}
 		// A object with x-kubernetes-preserve-unknown-fields but no properties or additionalProperties is treated
 		// as an empty object.
-		if schema.XPreserveUnknownFields {
+		if isXPreserveUnknownFields(schema) {
 			return &unstructuredMap{
 				value:  m,
 				schema: schema,
-				propSchema: func(key string) (*structuralschema.Structural, bool) {
+				propSchema: func(key string) (*spec.Schema, bool) {
 					return nil, false
 				},
 			}
@@ -99,7 +98,7 @@ func UnstructuredToVal(unstructured interface{}, schema *structuralschema.Struct
 		return types.NewErr("invalid object type, expected either Properties or AdditionalProperties with Allows=true and non-empty Schema")
 	}
 
-	if schema.Type == "array" {
+	if schema.Type.Contains("array") {
 		l, ok := unstructured.([]interface{})
 		if !ok {
 			return types.NewErr("invalid data, expected an array for the provided schema with type=array")
@@ -107,31 +106,30 @@ func UnstructuredToVal(unstructured interface{}, schema *structuralschema.Struct
 		if schema.Items == nil {
 			return types.NewErr("invalid array type, expected Items with a non-empty Schema")
 		}
-		typedList := unstructuredList{elements: l, itemsSchema: schema.Items}
-		listType := schema.XListType
-		if listType != nil {
-			switch *listType {
+		typedList := unstructuredList{elements: l, itemsSchema: schema.Items.Schema}
+		listType := getXListType(schema)
+		if listType != "" {
+			switch listType {
 			case "map":
-				mapKeys := schema.Extensions.XListMapKeys
+				mapKeys := getXListMapKeys(schema)
 				return &unstructuredMapList{unstructuredList: typedList, escapedKeyProps: escapeKeyProps(mapKeys)}
 			case "set":
 				return &unstructuredSetList{unstructuredList: typedList}
 			case "atomic":
 				return &typedList
 			default:
-				return types.NewErr("invalid x-kubernetes-list-type, expected 'map', 'set' or 'atomic' but got %s", *listType)
+				return types.NewErr("invalid x-kubernetes-list-type, expected 'map', 'set' or 'atomic' but got %s", listType)
 			}
 		}
 		return &typedList
 	}
 
-	if schema.Type == "string" {
+	if schema.Type.Contains("string") {
 		str, ok := unstructured.(string)
 		if !ok {
 			return types.NewErr("invalid data, expected string, got %T", unstructured)
 		}
-		if schema.ValueValidation != nil {
-			switch schema.ValueValidation.Format {
+		switch schema.Format {
 		case "duration":
 			d, err := strfmt.ParseDuration(str)
 			if err != nil {
@@ -158,10 +156,10 @@ func UnstructuredToVal(unstructured interface{}, schema *structuralschema.Struct
 			}
 			return types.Bytes(base64)
 		}
-		}
+
 		return types.String(str)
 	}
-	if schema.Type == "number" {
+	if schema.Type.Contains("number") {
 		switch v := unstructured.(type) {
 		// float representations of whole numbers (e.g. 1.0, 0.0) can convert to int representations (e.g. 1, 0) in yaml
 		// to json translation, and then get parsed as int64s
@@ -180,7 +178,7 @@ func UnstructuredToVal(unstructured interface{}, schema *structuralschema.Struct
 			return types.NewErr("invalid data, expected float, got %T", unstructured)
 		}
 	}
-	if schema.Type == "integer" {
+	if schema.Type.Contains("integer") {
 		switch v := unstructured.(type) {
 		case int:
 			return types.Int(v)
@@ -192,7 +190,7 @@ func UnstructuredToVal(unstructured interface{}, schema *structuralschema.Struct
 			return types.NewErr("invalid data, expected int, got %T", unstructured)
 		}
 	}
-	if schema.Type == "boolean" {
+	if schema.Type.Contains("boolean") {
 		b, ok := unstructured.(bool)
 		if !ok {
 			return types.NewErr("invalid data, expected bool, got %T", unstructured)
@@ -200,7 +198,7 @@ func UnstructuredToVal(unstructured interface{}, schema *structuralschema.Struct
 		return types.Bool(b)
 	}
 
-	if schema.XPreserveUnknownFields {
+	if isXPreserveUnknownFields(schema) {
 		return &unknownPreserved{u: unstructured}
 	}
 
@@ -420,7 +418,7 @@ func (t *unstructuredSetList) Add(other ref.Val) ref.Val {
 // unstructuredList represents an unstructured data instance of an OpenAPI array with x-kubernetes-list-type=atomic (the default).
 type unstructuredList struct {
 	elements    []interface{}
-	itemsSchema *structuralschema.Structural
+	itemsSchema *spec.Schema
 }
 
 var _ = traits.Lister(&unstructuredList{})
@@ -550,9 +548,9 @@ func (t *unstructuredList) Size() ref.Val {
 // unstructuredMap represented an unstructured data instance of an OpenAPI object.
 type unstructuredMap struct {
 	value  map[string]interface{}
-	schema *structuralschema.Structural
+	schema *spec.Schema
 	// propSchema finds the schema to use for a particular map key.
-	propSchema func(key string) (*structuralschema.Structural, bool)
+	propSchema func(key string) (*spec.Schema, bool)
 }
 
 var _ = traits.Mapper(&unstructuredMap{})

```